pkg/3scale/amp/operator/deployment_config_reconciler.go

```diff
@@ -4,228 +4,108 @@ import (
+	return &DeploymentConfigBaseReconciler{
+		BaseAPIManagerLogicReconciler: baseAPIManagerLogicReconciler,
+		reconciler:                    reconciler,
 	}
 }
 
-func (r *DeploymentConfigReconciler) Reconcile(desiredDeploymentConfig *appsv1.DeploymentConfig) error {
-	// We assume desired namespace and ownerref are set already
-	objectInfo := ObjectInfo(desiredDeploymentConfig)
-	existingDeploymentConfig := &appsv1.DeploymentConfig{}
-	err := r.Client().Get(context.TODO(), types.NamespacedName{Name: desiredDeploymentConfig.Name, Namespace: desiredDeploymentConfig.Namespace}, existingDeploymentConfig)
+func (r *DeploymentConfigBaseReconciler) Reconcile(desiredDC *appsv1.DeploymentConfig) error {
+	objectInfo := ObjectInfo(desiredDC)
+	existingDC := &appsv1.DeploymentConfig{}
+	err := r.Client().Get(
+		context.TODO(),
+		types.NamespacedName{Name: desiredDC.Name, Namespace: r.apiManager.GetNamespace()},
+		existingDC)
 	if err != nil {
 		if errors.IsNotFound(err) {
-			createErr := r.Client().Create(context.TODO(), desiredDeploymentConfig)
+			createErr := r.createResource(desiredDC)
 			if createErr != nil {
 				r.Logger().Error(createErr, fmt.Sprintf("Error creating object %s. Requeuing request...", objectInfo))
 				return createErr
 			}
-			r.Logger().Info(fmt.Sprintf("Created object %s", objectInfo))
 			return nil
 		}
 		return err
 	}
 
-	// Set in the desired some fields that are automatically set
-	// by Kubernetes controllers as defaults that are not defined in
-	// our logic
-	// TODO
-
-	// reconcile update
-	needsUpdate, err := r.ensureDeploymentConfig(existingDeploymentConfig, desiredDeploymentConfig)
-	if err != nil {
-		return err
-	}
-
-	if needsUpdate {
-		r.Logger().Info(fmt.Sprintf("Updating DeploymentConfig %s", objectInfo))
-		err := r.Client().Update(context.TODO(), existingDeploymentConfig)
-		if err != nil {
-			r.Logger().Error(err, fmt.Sprintf("error updating DeploymentConfig %s", objectInfo))
-			return err
-		}
+	if r.reconciler.IsUpdateNeeded(desiredDC, existingDC) {
+		return r.updateResource(existingDC)
 	}
 
 	return nil
 }
 
-func (r *DeploymentConfigReconciler) ensureDeploymentConfig(updated, desired *appsv1.DeploymentConfig) (bool, error) {
-	changed := false
+func DeploymentConfigReconcileContainerResources(desired, existing *appsv1.DeploymentConfig, logger logr.Logger) bool {
+	desiredName := ObjectInfo(desired)
+	update := false
 
-	objectMetaChanged, err := r.ObjectMetaMerger.EnsureObjectMeta(&updated.ObjectMeta, &desired.ObjectMeta)
-	if err != nil {
-		return false, err
-	}
-	if objectMetaChanged {
-		changed = true
-	}
-
-	// Reconcile PodTemplateSpec labels
-	if r.ObjectMetaMerger.EnsureLabels(updated.Spec.Template.ObjectMeta.Labels, desired.Spec.Template.ObjectMeta.Labels) {
-		changed = true
+	//
+	// Check container resource requirements
+	//
+	if len(desired.Spec.Template.Spec.Containers) != 1 {
```

Comment:
Is this check really needed?
It should never happen. If it's not 1 it means that the source code itself of the definition of the DeploymentConfig is wrong.
Containers is not the number of "replicas" defined but how many container definitions there are in the DeploymentConfig

Reply:
this is an `assert`. Makes sure about assumptions made, and panics with clear error message instead some `index out of range` confusing error.